core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/pipeline/stages/AsyncExecutionFailureExceptionReportingStage.java

```diff
@@ -58,19 +59,23 @@ public AsyncExecutionFailureExceptionReportingStage(RequestPipeline<SdkHttpFullR
     }
 
     /**
-     * Report the failure to the execution interceptors. Swallow any exceptions thrown from the interceptor since we don't
-     * want to replace the execution failure.
+     * Report the failure to the execution interceptors. Swallow any non-SdkServiceExceptions thrown from the interceptor since
+     * we don't want to replace the execution failure.
      *
      * @param context The execution context.
-     * @param failure     The execution failure.
+     * @param failure The execution failure.
      */
-    private static void reportFailureToInterceptors(RequestExecutionContext context, Throwable failure) {
+    private static Throwable reportFailureToInterceptors(RequestExecutionContext context, Throwable failure) {
         try {
             Context.FailedExecution failedContext =
                     new DefaultFailedExecutionContext(context.executionContext().interceptorContext(), failure);
             context.interceptorChain().onExecutionFailure(failedContext, context.executionAttributes());
+        } catch (SdkServiceException e) {
```

Comment:
Should we go all-out and propagate `SdkException`? All of the exceptions? Do we have similar behavior for the other on* methods, that suppresses exceptions that are thrown?

Reply:
The swallowing behavior is only `onExecuteFailure` since at this point there's already an `Exception` thrown higher up. The other `on*` methods let the exception propagate

https://github.com/aws/aws-sdk-java-v2/blob/fee061bf19d35435b42df3a53692757e037dc057/core/sdk-core/src/test/java/software/amazon/awssdk/core/client/AsyncClientHandlerInterceptorExceptionTest.java#L142-L160

Reply:
I'm fine with propagating `SdkException`. @shorea  Do you have any concerns?

Reply:
Should we add another interceptor method that lets you modify the exception so you don't have to throw?

Reply:
Andrew coming in with the BIG ideas. I like it!

Reply:
Added `modifyException` interceptor method.